src/data/finset.lean

```diff
@@ -2103,6 +2151,26 @@ end
 
 end max_min
 
+/-- An alternate version of `exists_max` with different typeclass requirements. -/
+lemma exists_max_of_decidable_range [linear_order α] [decidable_eq β] (s : finset β) (f : β → α) :
```

Comment:
I don't see why you want to have two statements: you can have a theorem assuming just `[linear_order α]` by going classical, using `classical.DLO`, and then copying your proof of `exists_max`. Since the statement is just an existence statement, going classical can not hurt.

Reply:
I wrote this for the proof [here](https://github.com/b-mehta/combinatorics/blob/7f39c2dd4e489bd10279c72bb9fcf0ee3024e78e/src/colex.lean#L116) - this statement was called `exists_max'` there. Using `exists_max`, the typeclass requirements would be `decidable_linear_order` instead of the weaker ones I used there.

Reply:
Sure. What I am saying is that you can have your proof of `exists_max` with the weaker typeclass assumption, with
```lean
lemma exists_max_image {α : Type*} [linear_order α] (s : finset β) (f : β → α) (h : s.nonempty) : 
  ∃ x ∈ s, ∀ x' ∈ s, f x' ≤ f x :=
begin
  letI := classical.DLO α,
  cases max_of_nonempty (h.image f) with y hy,
  rcases mem_image.mp (mem_of_max hy) with ⟨x, hx, rfl⟩,
  exact ⟨x, hx, λ x' hx', le_max_of_mem (mem_image_of_mem f hx') hy⟩,
end
```

Reply:
Oh hmm, I thought doing this would make the instance noncomputable but it doesn't seem to... Why is that?

Reply:
The conclusion of your lemma is a proposition, which has no computational content.

Reply:
I've removed `exists_max_of_decidable_range` and moved `exists_max` and `exists_min` so they use the weaker typeclass requirements as you suggest